File: bézier/julia/34-bezier-curvature.jl
# ============================================================
# Bezier Curve + Curvature Comb — Generic (2D or 3D)
# Julia port of 34-bezier-curvature.py
# Uses GLMakie (interactive, rotatable)
# ============================================================

using GLMakie
using Printf
using LinearAlgebra
using Statistics


# ============================================================
# -- Switch  (only change this one line) ---------------------
# ============================================================

const CASE = 2   # 1 = Quadratic 2D,  2 = Quartic 3D


# ============================================================
# -- Case Data  (do not change) ------------------------------
# ============================================================

const CASE_DATA = Dict(
    1 => (
        points      = Float64[1 1;
                               3 3;
                               5 1],
        u_highlight = [0.00, 0.25, 0.50, 0.75, 1.00],
        u_dc        = 0.50,
        output      = "34-bezier-curvature-quadratic-2d.png",
    ),
    2 => (
        points      = Float64[0 0 1;
                               0 4 1;
                               4 0 1;
                               4 4 1;
                               5 4 1],
        u_highlight = [0.00, 0.25, 0.50, 0.75, 1.00],
        u_dc        = 0.50,
        output      = "34-bezier-curvature-quartic-3d.png",
    ),
)


# ============================================================
# -- Bezier Computation --------------------------------------
# ============================================================

struct BezierCurve
    points  ::Matrix{Float64}   # (n × dim)
    n       ::Int
    degree  ::Int
    dim     ::Int
    # Derivative control point arrays
    d1_pts  ::Matrix{Float64}   # (n-1 × dim)  Q_i = n*(P_{i+1}-P_i)
    d2_pts  ::Union{Matrix{Float64}, Nothing}  # (n-2 × dim)
end

function BezierCurve(points::Matrix{Float64})
    n      = size(points, 1)
    deg    = n - 1
    dim    = size(points, 2)
    d1_pts = deg .* diff(points, dims=1)                    # (n-1 × dim)
    d2_pts = size(d1_pts, 1) > 1 ?
             (deg - 1) .* diff(d1_pts, dims=1) : nothing   # (n-2 × dim)
    BezierCurve(points, n, deg, dim, d1_pts, d2_pts)
end

function degree_name(bz::BezierCurve)
    names = Dict(2=>"Quadratic", 3=>"Cubic", 4=>"Quartic", 5=>"Quintic")
    get(names, bz.degree, "Degree-$(bz.degree)")
end

dim_name(bz::BezierCurve) = bz.dim == 3 ? "3D" : "2D"


# -- Bernstein evaluation (numeric, no symbolics) ------------

function bernstein(n::Int, i::Int, u::Float64)
    binomial(n, i) * u^i * (1 - u)^(n - i)
end

function eval_poly(pts::Matrix{Float64}, u::Float64)
    # Evaluate a Bezier curve defined by pts at parameter u
    n   = size(pts, 1) - 1
    sum(bernstein(n, i-1, u) .* pts[i,:] for i in 1:size(pts,1))
end

function bz_point(bz::BezierCurve, u::Float64)
    eval_poly(bz.points, u)
end

function bz_d1(bz::BezierCurve, u::Float64)
    eval_poly(bz.d1_pts, u)
end

function bz_d2(bz::BezierCurve, u::Float64)
    isnothing(bz.d2_pts) ? zeros(bz.dim) : eval_poly(bz.d2_pts, u)
end

function bz_curve(bz::BezierCurve, n_pts::Int=300)
    u_vals = LinRange(0.0, 1.0, n_pts)
    hcat([bz_point(bz, u) for u in u_vals]...)'   # (n_pts × dim)
end

function bz_highlight(bz::BezierCurve, u_list::Vector{Float64})
    hcat([bz_point(bz, u) for u in u_list]...)'
end

function bz_curvature(bz::BezierCurve, u::Float64)
    d1 = bz_d1(bz, u)
    d2 = bz_d2(bz, u)
    speed = norm(d1)
    speed < 1e-10 && return 0.0
    if bz.dim == 2
        abs(d1[1]*d2[2] - d1[2]*d2[1]) / speed^3
    else
        norm(cross(d1, d2)) / speed^3
    end
end

function de_casteljau(bz::BezierCurve, u::Float64)
    levels  = Matrix{Float64}[copy(bz.points)]
    current = copy(bz.points)
    while size(current, 1) > 1
        n = size(current, 1)
        current = vcat([((1-u) .* current[j,:] .+ u .* current[j+1,:])' for j in 1:n-1]...)
        push!(levels, current)
    end
    return levels
end


# ============================================================
# -- Colors --------------------------------------------------
# ============================================================

const COLOR_CURVE   = :steelblue
const COLOR_POLYGON = :crimson
const COLOR_CTRL    = :crimson
const COLOR_POINT   = :darkgreen
const COLOR_LABEL   = :purple
const COLOR_COMB    = RGBAf(0.565, 0.643, 0.682, 1.0)   # #90A4AE
const COLOR_DC      = [:darkorange, :deeppink, :darkviolet, :darkgreen]


# ============================================================
# -- Curvature Comb ------------------------------------------
# ============================================================

function draw_comb!(ax, bz::BezierCurve, u_vals, curve_pts,
                    kappas, comb_scale, is_3d)
    step  = max(1, length(u_vals) ÷ 60)
    u_sub = u_vals[1:step:end]
    m     = length(u_sub)

    base = hcat([bz_point(bz, u) for u in u_sub]...)'   # (m × dim)
    d1v  = hcat([bz_d1(bz, u)   for u in u_sub]...)'
    d2v  = hcat([bz_d2(bz, u)   for u in u_sub]...)'
    kap  = [bz_curvature(bz, u) for u in u_sub]

    tips = similar(base)

    if !is_3d
        # 2D: rotate tangent 90 degrees to get normal
        for i in 1:m
            spd = norm(d1v[i,:])
            spd = spd > 1e-10 ? spd : 1e-10
            T   = d1v[i,:] / spd
            N   = [-T[2], T[1]]
            tips[i,:] = base[i,:] .+ comb_scale * kap[i] .* N
        end
        # Spike lines
        for i in 1:m
            lines!(ax,
                [base[i,1], tips[i,1]],
                [base[i,2], tips[i,2]],
                color=(COLOR_COMB, 0.65), linewidth=1.2)
        end
        # Envelope
        lines!(ax, tips[:,1], tips[:,2],
            color=(COLOR_COMB, 0.5), linewidth=1.0,
            linestyle=:dash, label="Curvature comb")
    else
        # 3D: principal normal N = (P' × P'') × P'
        for i in 1:m
            d1i = d1v[i,:]
            d2i = d2v[i,:]
            N_raw = cross(cross(d1i, d2i), d1i)
            n_mag = norm(N_raw)
            N     = n_mag > 1e-10 ? N_raw / n_mag : zeros(3)
            tips[i,:] = base[i,:] .+ comb_scale * kap[i] .* N
        end
        # Spike lines
        for i in 1:m
            lines!(ax,
                [base[i,1], tips[i,1]],
                [base[i,2], tips[i,2]],
                [base[i,3], tips[i,3]],
                color=(COLOR_COMB, 0.65), linewidth=1.2)
        end
        # Envelope
        lines!(ax, tips[:,1], tips[:,2], tips[:,3],
            color=(COLOR_COMB, 0.5), linewidth=1.0,
            linestyle=:dash, label="Curvature comb")
    end
end


# ============================================================
# -- Plot ----------------------------------------------------
# ============================================================

function draw_and_show(
        bz         ::BezierCurve,
        u_highlight::Vector{Float64},
        u_dc       ::Float64,
        filename   ::String,
        n_pts      ::Int = 300)

    is_3d = bz.dim == 3
    title = "Bezier Curve + Curvature Comb: $(degree_name(bz)) " *
            "($(dim_name(bz)), $(bz.n) Control Points)"

    u_vals    = collect(LinRange(0.0, 1.0, n_pts))
    curve_pts = bz_curve(bz, n_pts)
    kappas    = [bz_curvature(bz, u) for u in u_vals]

    # Auto scale comb
    ext      = maximum(maximum(curve_pts[:,1:2], dims=1) .- minimum(curve_pts[:,1:2], dims=1))
    kap_95   = quantile(kappas, 0.95)
    comb_scale = kap_95 > 1e-10 ? 0.15 * ext / kap_95 : 0.1

    fig = Figure(size = (is_3d ? 900 : 800, is_3d ? 650 : 620),
                 backgroundcolor = :white)

    if is_3d
        ax = Axis3(fig[1, 1],
            title           = title,
            titlesize       = 13,
            titlegap        = 12,
            xlabel          = "x",
            ylabel          = "y",
            zlabel          = "z",
            xlabelsize      = 11,
            ylabelsize      = 11,
            zlabelsize      = 11,
            elevation       = deg2rad(35),
            azimuth         = deg2rad(295),
            limits          = (-0.5, 6.0, -0.5, 5.0, 0.0, 2.0),
            backgroundcolor = :white,
            xgridcolor      = (:lightgray, 0.5),
            ygridcolor      = (:lightgray, 0.5),
            zgridcolor      = (:lightgray, 0.5),
        )
    else
        ax = Axis(fig[1, 1],
            title           = title,
            titlesize       = 13,
            titlegap        = 12,
            xlabel          = "x",
            ylabel          = "y",
            xlabelsize      = 11,
            ylabelsize      = 11,
            backgroundcolor = :white,
            xgridcolor      = (:lightgray, 0.6),
            ygridcolor      = (:lightgray, 0.6),
        )
    end

    # ── Curvature comb (drawn first, sits under curve) ───────
    draw_comb!(ax, bz, u_vals, curve_pts, kappas, comb_scale, is_3d)

    # ── Bezier curve ─────────────────────────────────────────
    if is_3d
        lines!(ax, curve_pts[:,1], curve_pts[:,2], curve_pts[:,3],
            color=COLOR_CURVE, linewidth=2.5, label="Bezier Curve")
    else
        lines!(ax, curve_pts[:,1], curve_pts[:,2],
            color=COLOR_CURVE, linewidth=2.5, label="Bezier Curve")
    end

    # ── Control polygon ──────────────────────────────────────
    cp = bz.points
    if is_3d
        lines!(ax, cp[:,1], cp[:,2], cp[:,3],
            color=COLOR_POLYGON, linestyle=:dash, linewidth=1.5,
            label="Control Polygon")
        scatter!(ax, cp[:,1], cp[:,2], cp[:,3],
            color=COLOR_POLYGON, markersize=9, strokewidth=0)
    else
        lines!(ax, cp[:,1], cp[:,2],
            color=COLOR_POLYGON, linestyle=:dash, linewidth=1.5,
            label="Control Polygon")
        scatter!(ax, cp[:,1], cp[:,2],
            color=COLOR_POLYGON, markersize=9, strokewidth=0)
    end

    # ── Control point labels ──────────────────────────────────
    for i in 1:bz.n
        pt     = cp[i,:]
        coords = join([@sprintf("%.4g", v) for v in pt], ", ")
        name   = "P$(i-1)($coords)"
        dy     = i % 2 == 1 ? 0.15 : -0.28
        if is_3d
            text!(ax, pt[1]+0.10, pt[2]+dy, pt[3]+0.05,
                text=name, color=COLOR_CTRL, fontsize=8, font=:bold)
        else
            text!(ax, pt[1]+0.10, pt[2]+dy,
                text=name, color=COLOR_CTRL, fontsize=9, font=:bold)
        end
    end

    # ── Highlighted points ────────────────────────────────────
    hpts = bz_highlight(bz, u_highlight)
    if is_3d
        scatter!(ax, hpts[:,1], hpts[:,2], hpts[:,3],
            color=COLOR_POINT, markersize=7, strokewidth=0,
            label="Points at u")
        for (i, u) in enumerate(u_highlight)
            pt = hpts[i,:]
            text!(ax, pt[1]+0.08, pt[2]+0.08, pt[3]+0.05,
                text=@sprintf("u=%.2f", u),
                color=COLOR_LABEL, fontsize=7)
        end
    else
        scatter!(ax, hpts[:,1], hpts[:,2],
            color=COLOR_POINT, markersize=9, strokewidth=0,
            label="Points at u")
        for (i, u) in enumerate(u_highlight)
            pt     = hpts[i,:]
            coords = join([@sprintf("%.3g", v) for v in pt], ", ")
            lbl    = @sprintf("u=%.2f\n(%s)", u, coords)
            dx     = u == 1.0 ? -0.55 : 0.08
            dy     = u in (0.0, 1.0) ? -0.28 : 0.10
            text!(ax, pt[1]+dx, pt[2]+dy,
                text=lbl, color=COLOR_LABEL, fontsize=8)
        end
    end

    # ── De Casteljau levels ───────────────────────────────────
    levels = de_casteljau(bz, u_dc)
    for lv in 2:length(levels)
        pts      = levels[lv]
        col      = COLOR_DC[min(lv-1, length(COLOR_DC))]
        is_final = (lv == length(levels))
        lbl      = is_final ? @sprintf("DC final u=%.2f", u_dc) : "DC level $(lv-1)"

        if !is_final
            if is_3d
                lines!(ax, pts[:,1], pts[:,2], pts[:,3],
                    color=col, linewidth=1.4, label=lbl)
            else
                lines!(ax, pts[:,1], pts[:,2],
                    color=col, linewidth=1.4, label=lbl)
            end
        end

        if is_3d
            scatter!(ax, pts[:,1], pts[:,2], pts[:,3],
                color=col, markersize=is_final ? 12 : 7,
                marker=is_final ? :diamond : :circle, strokewidth=0,
                label=is_final ? lbl : "")
        else
            scatter!(ax, pts[:,1], pts[:,2],
                color=col, markersize=is_final ? 12 : 7,
                marker=is_final ? :diamond : :circle, strokewidth=0,
                label=is_final ? lbl : "")
        end

        prev = levels[lv-1]
        for j in 1:size(pts, 1)
            for k in [j, j+1]
                if is_3d
                    lines!(ax,
                        [prev[k,1], pts[j,1]],
                        [prev[k,2], pts[j,2]],
                        [prev[k,3], pts[j,3]],
                        color=(col, 0.35), linewidth=0.8, linestyle=:dot)
                else
                    lines!(ax,
                        [prev[k,1], pts[j,1]],
                        [prev[k,2], pts[j,2]],
                        color=(col, 0.35), linewidth=0.8, linestyle=:dot)
                end
            end
        end
    end

    # ── Equation box ─────────────────────────────────────────
    n     = bz.degree
    terms = join(["B($n,$(i-1))*P$(i-1)" for i in 1:bz.n], " + ")
    eq    = "P(u) = $terms\n" *
            "B(n,i,u) = C(n,i)*u^i*(1-u)^(n-i)\n" *
            "kappa = |P' x P''| / |P'|^3"
    Label(fig[0, 1],
        eq,
        fontsize  = 8,
        font      = "Courier New",
        halign    = :left,
        padding   = (8, 8, 4, 4),
        tellwidth = false,
    )

    # ── Legend ───────────────────────────────────────────────
    axislegend(ax,
        position        = is_3d ? :lt : :rb,
        labelsize       = 8,
        framecolor      = (:black, 0.3),
        backgroundcolor = (:white, 0.8),
    )

    # ── 2D auto limits ───────────────────────────────────────
    if !is_3d
        all_x = vcat(cp[:,1], curve_pts[:,1])
        all_y = vcat(cp[:,2], curve_pts[:,2])
        m = 0.8
        xlims!(ax, minimum(all_x)-m, maximum(all_x)+m)
        ylims!(ax, minimum(all_y)-m, maximum(all_y)+m)
    end

    save(filename, fig, px_per_unit = 1.5)
    println("Saved: $filename")

    display(fig)
    println("Press Enter to close...")
    readline()
end


# ============================================================
# -- Entry Point ---------------------------------------------
# ============================================================

function main()
    data = CASE_DATA[CASE]
    bz   = BezierCurve(data.points)
    draw_and_show(bz, data.u_highlight, data.u_dc, data.output)
end

main()

File: bézier/python/34-bezier-curvature.py
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.interpolate import BPoly


# ============================================================
# -- Switch  (only change this one line) ---------------------
# ============================================================

CASE = 2   # 1 = Quadratic 2D,  2 = Quartic 3D


# ============================================================
# -- CASE Data  (do not change) ------------------------------
# ============================================================

CASE_DATA = {
    1: dict(
        points      = np.array([
            [1, 1],   # P0
            [3, 3],   # P1
            [5, 1],   # P2
        ]),
        u_highlight = [0.00, 0.25, 0.50, 0.75, 1.00],
        u_dc        = 0.50,
        output      = '34-bezier_curvature_quadratic_2d.png',
    ),
    2: dict(
        points      = np.array([
            [0, 0, 1],   # P0
            [0, 4, 1],   # P1
            [4, 0, 1],   # P2
            [4, 4, 1],   # P3
            [5, 4, 1],   # P4
        ]),
        u_highlight = [0.00, 0.25, 0.50, 0.75, 1.00],
        u_dc        = 0.50,
        output      = '34-bezier_curvature_quartic_3d.png',
    ),
}


# ============================================================
# -- BezierCurve  (generic) ----------------------------------
# ============================================================

class BezierCurve:
    """
    Generic Bezier curve, any degree, 2D or 3D.

    Curve evaluation uses scipy.interpolate.BPoly.

    Curvature uses derivative Bezier curves:
      P'(u)  — BPoly with control points Q_i = n*(P_{i+1} - P_i)
      P''(u) — BPoly with control points R_i = (n-1)*(Q_{i+1} - Q_i)

    kappa = |P' x P''| / |P'|^3
      2D: cross product is scalar  x'y'' - y'x''
      3D: cross product is a vector, take its norm
    """

    def __init__(self, points: np.ndarray):
        self.points = points
        self.n      = len(points)
        self.degree = self.n - 1
        self.dim    = points.shape[1]

        c = points[:, np.newaxis, :]
        self._bpoly = BPoly(c, [0, 1])

        Q = self.degree * np.diff(points, axis=0)
        self._bpoly_d1 = BPoly(Q[:, np.newaxis, :], [0, 1])

        if len(Q) > 1:
            R = (self.degree - 1) * np.diff(Q, axis=0)
            self._bpoly_d2 = BPoly(R[:, np.newaxis, :], [0, 1])
        else:
            self._bpoly_d2 = None

        print(c)
        print(Q)
        print(R)

    def point(self, u):
        return self._bpoly(u)

    def curve(self, n_pts=300):
        return self._bpoly(np.linspace(0, 1, n_pts))

    def highlight_points(self, u_list):
        return self._bpoly(np.array(u_list))

    def de_casteljau(self, u):
        levels  = [self.points.copy()]
        current = self.points.copy()
        while len(current) > 1:
            current = np.array([
                (1-u)*current[j] + u*current[j+1]
                for j in range(len(current) - 1)
            ])
            levels.append(current)
        return levels

    def d1(self, u):
        return self._bpoly_d1(u)

    def d2(self, u):
        if self._bpoly_d2 is None:
            return np.zeros((len(np.atleast_1d(u)), self.dim))
        return self._bpoly_d2(u)

    def curvature(self, u):
        u_arr = np.atleast_1d(u)
        d1    = self.d1(u_arr)
        d2    = self.d2(u_arr)
        if self.dim == 2:
            cross = np.abs(d1[:, 0]*d2[:, 1] - d1[:, 1]*d2[:, 0])
        else:
            cross = np.linalg.norm(np.cross(d1, d2), axis=1)
        speed = np.linalg.norm(d1, axis=1)
        denom = np.where(speed > 1e-10, speed**3, np.inf)
        return cross / denom

    @property
    def degree_name(self):
        names = {2:'Quadratic', 3:'Cubic', 4:'Quartic', 5:'Quintic'}
        return names.get(self.degree, f'Degree-{self.degree}')

    @property
    def dim_name(self):
        return '3D' if self.dim == 3 else '2D'


# ============================================================
# -- BezierPlot  (full plot + curvature comb) ----------------
# ============================================================

class BezierPlot:
    """
    Full plot matching 32-bezier-scipy.py output:
      curve, control polygon, highlight points, DC levels,
      equation box — plus curvature comb added on top.
    """

    COLOR_CURVE   = '#1565C0'
    COLOR_POLYGON = '#E53935'
    COLOR_CTRL    = '#E53935'
    COLOR_POINT   = '#1B5E20'
    COLOR_LABEL   = '#4A148C'
    COLOR_COMB    = '#90A4AE'
    COLOR_DC = [
        '#F57F17',   # level 1 — amber
        '#EC407A',   # level 2 — pink
        '#7B1FA2',   # level 3 — violet
        '#1B5E20',   # level 4 — green (final point)
    ]

    def __init__(self, bezier, u_highlight, u_dc=0.5,
                 n_pts=300, comb_scale=None):
        self.bezier      = bezier
        self.u_highlight = u_highlight
        self.u_dc        = u_dc
        self.n_pts       = n_pts
        self.is_3d       = (bezier.dim == 3)

        self.u_vals    = np.linspace(0, 1, n_pts)
        self.curve_pts = bezier.curve(n_pts)
        self.kappa     = bezier.curvature(self.u_vals)

        if comb_scale is None:
            ext     = np.ptp(self.curve_pts[:, :2], axis=0).max()
            kap_max = np.percentile(self.kappa, 95)
            self.comb_scale = (0.15 * ext / kap_max
                               if kap_max > 1e-10 else 0.1)
        else:
            self.comb_scale = comb_scale

        sns.set_theme(style='whitegrid')
        if self.is_3d:
            self.fig = plt.figure(figsize=(10, 7))
            self.ax  = self.fig.add_subplot(111, projection='3d')
        else:
            self.fig, self.ax = plt.subplots(figsize=(9, 7))

    # -- Coordinate helpers ----------------------------------

    def _coords(self, pts):
        if self.is_3d:
            return pts[:, 0], pts[:, 1], pts[:, 2]
        return pts[:, 0], pts[:, 1]

    def _plot_line(self, pts, **kwargs):
        self.ax.plot(*self._coords(pts), **kwargs)

    def _scatter_pts(self, pts, **kwargs):
        if self.is_3d:
            xs, ys, zs = self._coords(pts)
            self.ax.scatter(xs, ys, zs, **kwargs)
        else:
            xs, ys = self._coords(pts)
            self.ax.scatter(xs, ys, **kwargs)

    def _annotate(self, pt, text, dx, dy, dz, color):
        if self.is_3d:
            self.ax.text(pt[0]+dx, pt[1]+dy, pt[2]+dz,
                         text, fontsize=8, color=color,
                         fontweight='bold')
        else:
            self.ax.annotate(text,
                             xy=(pt[0], pt[1]),
                             xytext=(pt[0]+dx, pt[1]+dy),
                             fontsize=8.5, color=color,
                             fontweight='bold')

    # -- Original 32-bezier-scipy drawing methods ------------

    def _draw_curve(self):
        pts = self.bezier.curve()
        self._plot_line(pts, color=self.COLOR_CURVE, linewidth=2.5,
                        label='Bezier Curve', zorder=3)

    def _draw_control_polygon(self):
        self._plot_line(self.bezier.points,
                        color=self.COLOR_POLYGON, linestyle='--',
                        linewidth=1.5, marker='o', markersize=8,
                        label='Control Polygon', zorder=4)

    def _label_control_points(self):
        for i, pt in enumerate(self.bezier.points):
            coords = ', '.join(f'{v:.4g}' for v in pt)
            name   = f'P{i}({coords})'
            dy     = 0.15 if i % 2 == 0 else -0.28
            self._annotate(pt, name, dx=0.10, dy=dy, dz=0.05,
                           color=self.COLOR_CTRL)

    def _draw_highlight_points(self):
        pts = self.bezier.highlight_points(self.u_highlight)
        self._scatter_pts(pts, color=self.COLOR_POINT,
                          s=60, zorder=5, label='Points at u')
        for u, pt in zip(self.u_highlight, pts):
            coords = ', '.join(f'{v:.3g}' for v in pt)
            dx = -0.55 if u == 1.0 else 0.08
            dy = -0.28 if u in (0.0, 1.0) else 0.10
            if self.is_3d:
                self.ax.text(pt[0]+0.08, pt[1]+0.08, pt[2]+0.05,
                             f'u={u:.2f}', fontsize=7,
                             color=self.COLOR_LABEL)
            else:
                self.ax.annotate(
                    f'u={u:.2f}\n({coords})',
                    xy=(pt[0], pt[1]),
                    xytext=(pt[0]+dx, pt[1]+dy),
                    fontsize=7.5, color=self.COLOR_LABEL,
                    bbox=dict(boxstyle='round,pad=0.2',
                              fc='white', alpha=0.65, ec='none'),
                )

    def _draw_de_casteljau(self):
        levels = self.bezier.de_casteljau(self.u_dc)
        for lv in range(1, len(levels) - 1):
            pts = levels[lv]
            col = self.COLOR_DC[min(lv-1, len(self.COLOR_DC)-1)]
            self._plot_line(pts, color=col, linewidth=1.4,
                            marker='o', markersize=6, zorder=6,
                            label=f'DC level {lv}')
        final = levels[-1]
        col   = self.COLOR_DC[min(len(levels)-2, len(self.COLOR_DC)-1)]
        self._scatter_pts(final, color=col, marker='D',
                          s=90, zorder=7,
                          label=f'DC final  u={self.u_dc:.2f}')

    def _add_equation_box(self):
        n     = self.bezier.degree
        terms = ' + '.join(
            f'B({n},{i})\u00b7P{i}' for i in range(self.bezier.n)
        )
        text = (f'P(u) = {terms}\n'
                f'B(n,i,u) = C(n,i)\u00b7u\u1d35\u00b7(1-u)^(n-i)\n'
                f'[evaluated via scipy.interpolate.BPoly]')
        if self.is_3d:
            self.fig.text(0.13, 0.92, text, fontsize=8,
                          fontfamily='monospace', va='top',
                          bbox=dict(boxstyle='round,pad=0.4',
                                    facecolor='lightyellow', alpha=0.85))
        else:
            self.ax.text(0.02, 0.97, text,
                         transform=self.ax.transAxes,
                         fontsize=8, fontfamily='monospace', va='top',
                         bbox=dict(boxstyle='round,pad=0.4',
                                   facecolor='lightyellow', alpha=0.85))

    # -- Curvature comb (new) --------------------------------

    def _draw_comb(self):
        """
        Perpendicular spikes along the curve,
        length proportional to kappa at each point.

        2D: normal = unit tangent rotated 90 degrees.
        3D: principal normal via N = (P' x P'') x P',
            normalized — points toward centre of curvature
            in full 3D space. Spikes drawn as 3D lines.
        """
        step  = max(1, self.n_pts // 60)
        u_sub = self.u_vals[::step]

        d1v   = self.bezier.d1(u_sub)   # (m, dim)
        d2v   = self.bezier.d2(u_sub)   # (m, dim)
        base  = self.bezier._bpoly(u_sub)  # (m, dim)
        kap_s = self.bezier.curvature(u_sub)

        if not self.is_3d:
            # 2D: rotate tangent 90 degrees
            speed = np.linalg.norm(d1v, axis=1, keepdims=True)
            speed = np.where(speed > 1e-10, speed, 1e-10)
            T = d1v / speed
            N = np.stack([-T[:, 1], T[:, 0]], axis=1)

            tip = base + self.comb_scale * kap_s[:, np.newaxis] * N

            for b, t in zip(base, tip):
                self.ax.plot([b[0], t[0]], [b[1], t[1]],
                             color=self.COLOR_COMB, alpha=0.65,
                             linewidth=1.2, zorder=2)
            self.ax.plot(tip[:, 0], tip[:, 1],
                         color=self.COLOR_COMB, alpha=0.5,
                         linewidth=1.0, linestyle='--', zorder=2,
                         label='Curvature comb')
        else:
            # 3D: principal normal N = (P' x P'') x P'
            cross_d1_d2 = np.cross(d1v, d2v)          # binormal direction
            N_raw       = np.cross(cross_d1_d2, d1v)  # principal normal

            norm = np.linalg.norm(N_raw, axis=1, keepdims=True)
            norm = np.where(norm > 1e-10, norm, 1e-10)
            N    = N_raw / norm

            tip = base + self.comb_scale * kap_s[:, np.newaxis] * N

            for b, t in zip(base, tip):
                self.ax.plot([b[0], t[0]], [b[1], t[1]], [b[2], t[2]],
                             color=self.COLOR_COMB, alpha=0.65,
                             linewidth=1.2)
            self.ax.plot(tip[:, 0], tip[:, 1], tip[:, 2],
                         color=self.COLOR_COMB, alpha=0.5,
                         linewidth=1.0, linestyle='--',
                         label='Curvature comb')

    # -- Decorations -----------------------------------------

    def _set_decorations(self):
        title = (f'Bezier Curve + Curvature Comb: '
                 f'{self.bezier.degree_name} '
                 f'({self.bezier.dim_name}, '
                 f'{self.bezier.n} Control Points)')
        self.ax.set_title(title, fontsize=13,
                          fontweight='bold', pad=14)
        self.ax.set_xlabel('x', fontsize=11)
        self.ax.set_ylabel('y', fontsize=11)
        if self.is_3d:
            self.ax.set_zlabel('z', fontsize=11)
        self.ax.legend(fontsize=8, loc='lower right')

        if not self.is_3d:
            pts   = self.bezier.points
            curve = self.bezier.curve()
            all_x = np.concatenate([pts[:, 0], curve[:, 0]])
            all_y = np.concatenate([pts[:, 1], curve[:, 1]])
            m = 0.8
            self.ax.set_xlim(all_x.min()-m, all_x.max()+m)
            self.ax.set_ylim(all_y.min()-m, all_y.max()+m)

    # -- Main ------------------------------------------------

    def draw(self):
        self._draw_comb()             # draw first so curve sits on top
        self._draw_curve()
        self._draw_control_polygon()
        self._label_control_points()
        self._draw_highlight_points()
        self._draw_de_casteljau()
        self._add_equation_box()
        self._set_decorations()

    def save(self, filename):
        self.fig.tight_layout()
        self.fig.savefig(filename, dpi=150)
        print(f'Saved: {filename}')

    def show(self):
        self.fig.tight_layout()
        plt.show()


# ============================================================
# -- Entry Point ---------------------------------------------
# ============================================================

def main() -> int:
    data   = CASE_DATA[CASE]
    bezier = BezierCurve(data['points'])
    plot   = BezierPlot(
        bezier,
        u_highlight = data['u_highlight'],
        u_dc        = data['u_dc'],
    )
    plot.draw()
    plot.save(data['output'])
    plot.show()
    return 0


if __name__ == '__main__':
    raise SystemExit(main())
